KEYCLOAK-14138 Upgrade OTP login screen

Pepo48 opened the pull request:
* edited related css and ftl theme resources
* fixed affected tests

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

@mabartos @vmuzikar @ssilvert PTAL once you got a chance. I'll attach the pipeline test run results later. Thanks!

Pepo48 committed:
KEYCLOAK-14138 Upgrade OTP login screen

* edited related css and ftl theme resources
* added tile component
* fixed IE11 compatibility
* fixed affected tests

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

mabartos commented on themes/src/main/resources/theme/keycloak/login/theme.properties:
```diff
@@ -2,7 +2,7 @@ parent=base
 
 styles=css/login.css
-stylesCommon=node_modules/patternfly/dist/css/patternfly.min.css node_modules/patternfly/dist/css/patternfly-additions.min.css lib/zocial/zocial.css
+stylesCommon=node_modules/patternfly/dist/css/patternfly.min.css web_modules/@patternfly/react-core/dist/styles/app.css web_modules/@patternfly/react-core/dist/styles/base.css node_modules/patternfly/dist/css/patternfly-additions.min.css lib/patternfly4/tile.css
```
I have some issues with margin and padding. If you add the `css` file in this order, like below, it should resolve that.
`
base.css, app.css, patternfly.min.css, patternfly-additions.min.css tile.css
`

vmuzikar replied:
@mabartos What do you mean? I haven't noticed any issues. :)

Pepo48 replied:
@vmuzikar (cc: @mabartos) me neither, anyway I change the order as suggested.

mabartos commented on themes/src/main/resources/theme/keycloak/common/resources/lib/patternfly4/tile.css:
```diff
@@ -0,0 +1,100 @@
+.pf-c-tile {
```
@Pepo48 Have you tried this on IE? IE doesn't support `var()` function. The basic screens should be available even in IE, as it was discussed [here](https://github.com/keycloak/keycloak/pull/7470#discussion_r503393541). So, there should be, at least, some similar appearance like with PF4.

Pepo48 replied:
@mabartos (cc: @vmuzikar ) I completely revamped the css styles of the tile component. Now the screen is IE11 compatible:
![obrázok](https://user-images.githubusercontent.com/12138171/97352393-ffc00c00-1892-11eb-94fe-f7e7d5439ef9.png)

mabartos commented on themes/src/main/resources/theme/base/login/login-otp.ftl:
```diff
@@ -9,13 +9,13 @@
-                                            </h2>
+                                        <div class="${properties.kcLoginOTPListItemHeaderClass!}">
+                                            <div class="${properties.kcLoginOTPListItemIconClass!}">
+                                              <i class="fa fa-mobile" aria-hidden="true"></i>
```
I think, it'd be better use property for the icon instead of hardcoded value. Then, customer can easily change the icon in `theme.properties`. WDYT @Pepo48 ?

Pepo48 replied:
@mabartos (cc: @vmuzikar ) fixed

mabartos commented on themes/src/main/resources/theme/keycloak/login/resources/css/login.css:
```diff
@@ -515,8 +515,12 @@ a.zocial  {
-.login-pf-page .card-pf{
-  margin-bottom: 10px;
+.otp-tile {
+    padding: 1.5rem;
```
Is it possible to add any `margin-bottom`? If user have defined f.e. 3 OTP credentials with longer name, there are two rows, but there are no space between them.

Pepo48 replied:
@mabartos (cc: @vmuzikar ) fixed for IE (see the screenshot above) and other browsers:
![obrázok](https://user-images.githubusercontent.com/12138171/97352769-5e858580-1893-11eb-98e4-747aa9c87850.png)

mabartos reviewed (commented):
@Pepo48 Thanks for your PR! It looks nice. I've added only some minor suggestions to consideration for you.

vmuzikar commented on themes/src/main/resources/theme/base/login/login-otp.ftl:
```diff
@@ -51,16 +51,16 @@
-                $('.card-pf-view-single-select').click(function() {
-                  if ($(this).hasClass('active'))
-                  { $(this).removeClass('active'); $(this).children().removeAttr('name'); }
+                $('.otp-tile').click(function() {
```
This is a bit off topic for this PR as we used JS here as well (but I just noticed it now :)). Didn't we have some policy not to use any JS on login pages where possible (of course e.g. webauthn wouldn't work without JS). @ssilvert?

mposolda replied:
As you pointed, we have JS already, so no regression of this PR :-) But we have follow-up JIRA for remove JS from login pages.

vmuzikar replied:
Ah, sorry – I forgot about [KEYCLOAK-12823](https://issues.redhat.com/browse/KEYCLOAK-12823) (even though I was the one who created it :)). On the other hand, that JIRA covers the authenticator selector page, not OTP page. Or did you mean some other JIRA.

mposolda replied:
Nope, I meant this one and forgot that it is specific only to authenticator selector page :-) Maybe we can slightly rename the existing JIRA and/or update it's description to cover both screens? Or create new JIRA?

vmuzikar replied:
I've updated KEYCLOAK-12823.

vmuzikar reviewed (commented):
Except for @mabartos mentioned, I haven't found any other issues. Thank you @Pepo48!

The failed test in Travis is probably unrelated to this. A rebase might fix it.

Pepo48 commented:
> Pepo48 force-pushed the Pepo48:KEYCLOAK-14138 branch from `5ada513` to `6b584a0`

Conflicts resolved.

mabartos reviewed (commented):
@Pepo48 LGTM. Well done!
Please, execute the UI tests, then I'll approve it. Thanks.

![adminOTP](https://user-images.githubusercontent.com/38039883/97695376-3b7df000-1aa4-11eb-8814-a6f77606f5a5.png)

mabartos reviewed (dismissed):
There are some failures in UI tests, but it's not caused by this PR, so these tests were executed successfully for this PR.
Thanks @Pepo48 !

vmuzikar commented on themes/src/main/resources/theme/keycloak/common/resources/lib/patternfly4/tile.css:
```diff
@@ -0,0 +1,202 @@
+/*Internet Explorer 11 compatibility workaround - IE does not support CSS variables */
```
If I understand it correctly, this directory is a 3rd party library. I'm not sure we should be adding custom files in there. It might even get deleted on some dependencies update. Or am I missing something? @ssilvert

Pepo48 replied:
@vmuzikar it isn't from a 3rd party library, it's actually a part of the most recent releases of PF4 [1][2] and it's just copied from there - purely just for the sake of the transition phase. 

This file can be deleted as soon as `@patternfly` module will be updated in the project. 

[1] https://www.patternfly.org/v4/components/tile
[2] https://github.com/patternfly/patternfly/blob/84ae888cf912693b209ac0cf8f63b3ab3ddf59c2/src/patternfly/components/Tile/tile.scss

Pepo48 replied:
edit: sry @vmuzikar, I misunderstood your input at first, to clarify that - it's just partially custom, the vast majority of the code comes from PF4 [1] and the IE optimization part is actually the only custom piece of code. In other words, we could eventually move just the IE workaround part from there (e.g. to login.css). But just a suggestion ;).

[1] https://github.com/patternfly/patternfly/blob/84ae888cf912693b209ac0cf8f63b3ab3ddf59c2/src/patternfly/components/Tile/tile.scss

[]

vmuzikar replied:
Thanks for the clarification. I still see it as a custom CSS file (manually added, edited, not managed by NPM, etc.) which as such shouldn't be placed among other (managed) dependencies. But let's wait for @ssilvert, maybe I'm wrong.

vmuzikar replied:
Btw, why not to put these changes into `login.css` alongside other IE hacks?

vmuzikar reviewed (commented):
Apart from my last comment looks good – even in IE! :)

Pepo48 commented:
I'm attaching the UI tests result: 
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2261/

Pepo48 commented:
I'm attaching the base tests result:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2285/

There are some seemingly related failures, but I was not able to reproduce them locally - I ran both broker and forms testsuite submodules (over 500 tests) without failure. Moreover BrowserFlowTest was specifically fixed in this PR.

Any thoughts @vmuzikar? Thanks!

Pepo48 commented:
UPDATE
I'm attaching the final base tests result (my yesterday's assumption was right - without failures):
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2288/

mposolda commented:
@Pepo48 I am sorry, but it seems this PR was somehow forgotten. Is it still valid? If yes, can you please rebase and run the admin console tests again to make sure there is no additional regression added in the meantime with this PR in combination with the other PRs merged in the meantime to Keycloak master (mostly those by @mabartos) ? Thanks!

Pepo48 commented:
> @Pepo48 I am sorry, but it seems this PR was somehow forgotten. Is it still valid? If yes, can you please rebase and run the admin console tests again to make sure there is no additional regression added in the meantime with this PR in combination with the other PRs merged in the meantime to Keycloak master (mostly those by @mabartos) ? Thanks!

@mposolda Hi Marek, yes it's still valid - I've just rebased it and run the UI tests again - I'll attach the results once they are finished. 
One discussion point still remains opened though - whether the tile.css should be placed in lib directory or not? From my perspective it's not a big deal and initially @ssilvert had approved this approach but later on @vmuzikar raised some concerns (mentioned above).

mabartos commented:
@Pepo48 Thanks for update. As @vmuzikar mentioned above, the custom `tile.css` should not be included in the directory with external dependencies. I'd prefer move that to the another dir as well. At this moment, the tiles are used only in the login theme and IMHO, I'd move it there. @Pepo48 WDYT?

Pepo48 commented:
> can you please rebase and run the admin console tests again

Here are the UI tests results - there are some failures but I don't see anything related right now:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2337/

> @Pepo48 Thanks for update. As @vmuzikar mentioned above, the custom `tile.css` should not be included in the directory with external dependencies. I'd prefer move that to the another dir as well. At this moment, the tiles are used only in the login theme and IMHO, I'd move it there. @Pepo48 WDYT?

As soon as both of you are fine with this approach, I can easily move it. @vmuzikar let us know the final decision :)

vmuzikar commented:
> As soon as both of you are fine with this approach, I can easily move it. @vmuzikar let us know the final decision :)

I say let's move it. :)

Pepo48 commented:
> I say let's move it. :)

I moved it and also adjusted the link in theme.properties accordingly:
https://github.com/keycloak/keycloak/blob/a9e35cd156c5e40e88b80e54be795f7835871b23/themes/src/main/resources/theme/keycloak/login/resources/css/tile.css#L1

https://github.com/keycloak/keycloak/blob/a9e35cd156c5e40e88b80e54be795f7835871b23/themes/src/main/resources/theme/keycloak/login/theme.properties#L4

vmuzikar reviewed (approved):
Thank you @Pepo48!

Pepo48 commented:
I'm attaching the results once again:
* base tests:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2344/
* UI tests:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/2342/

mabartos reviewed (approved):
Thanks @Pepo48 !

mposolda reviewed (approved):
@Pepo48 @mabartos @vmuzikar Thanks everyone!
